File: Data Structures & Algorithms/serialize-and-deserialize-binary-tree/submission-1.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Codec:
    
    # Encodes a tree to a single string.
    def serialize(self, root: Optional[TreeNode]) -> str:
        if not root: return ""
        res = []
        q = deque()
        q.append(root)
        while q:
            curr = q.popleft()
            if curr:
                res.append(str(curr.val))
                q.append(curr.left)
                q.append(curr.right)
            else:
                res.append('#')
        return ",".join(res)
        
    # Decodes your encoded data to tree.
    def deserialize(self, data: str) -> Optional[TreeNode]:
        if not data: return None
        
        data = data.split(",")

        root = TreeNode(data[0])
        q = deque()
        q.append(root)
        i = 1
        while q and i < len(data):
            curr = q.popleft()
            if curr is None: continue
            if data[i] != '#':
                curr.left = TreeNode(data[i])
                q.append(curr.left)
            else:
                curr.left = None
            
            if data[i + 1] != '#':
                curr.right = TreeNode(data[i + 1])
                q.append(curr.right)
            else:
                curr.right = None
            i += 2
        return root

File: Data Structures & Algorithms/lowest-common-ancestor-in-binary-search-tree/submission-1.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    TreeNode* lowestCommonAncestor(TreeNode* root, TreeNode* p, TreeNode* q) {
        if (p->val > q->val) swap(p, q);
        return dfs(root, p, q);
    }

    TreeNode* dfs(TreeNode* root, TreeNode* p, TreeNode* q) {
        if (!root) return nullptr;

        if (p->val <= root->val && root->val <= q->val) return root;
        if (q->val < root->val) return dfs(root->left, p, q);
        else return dfs(root->right, p, q);
    }
};
